sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,287 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/stdasm.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * @file
+ * @brief Control and Status Register (CSR) interface.
+ *
+ * A set of macros that provide both read and modify operations on CSRs.
+ * Compiling translation units that include this header file with `-DMOCK_CSR`
+ * will result in the CSR operations being replaced with a mocked
+ * implementation.
+ */
+
+/**
+ * Define the implementation macros.
+ *
+ * The implementation used depends on whether the CSR library is providing a
+ * real or a mocked interface.
+ */
+#ifdef MOCK_CSR
+
+uint32_t mock_csr_read(uint32_t addr);
+
+#define CSR_READ_IMPL(csr, dst)  \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
+    *dst = mock_csr_read(csr);   \
+  } while (0);
+
+void mock_csr_write(uint32_t addr, uint32_t value);
```

Comment:
I'm confused, shouldn't this return the old value...?

Reply:
An older iteration of this PR did exactly that. It used generated functions with a return value that could just be discarded if not needed. With macros the return values need to be passed as parameters so I think it makes sense to add separate `CSR_READ_WRITE`, `CSR_READ_SET_BITS` and `CSR_READ_CLEAR_BITS` macros along with their equivalent mocked implementations as/when we need them.

Reply:
We may change our minds depending on how many such macros we will need and how verbose the code becomes but SGTM for now.